File: python/vapor_idx/indexes/exact.py
# =============================================================================
# vapor-idx — indexes/exact.py
# Equality index: field → normalised_value → set[record_id]
# Supports: eq, neq, in, notIn
# =============================================================================

from __future__ import annotations
from collections import defaultdict
from typing import Any


class ExactIndex:
    def __init__(self) -> None:
        # field → normalised_value → set[id]
        self._index: dict[str, dict[str, set[str]]] = defaultdict(lambda: defaultdict(set))

    # ── Mutation ───────────────────────────────────────────────────────────────

    def add(self, field: str, value: Any, record_id: str) -> None:
        if value is None:
            return
        values = value if isinstance(value, list) else [value]
        for v in values:
            self._index[field][_normalise(v)].add(record_id)

    def remove(self, field: str, value: Any, record_id: str) -> None:
        if value is None:
            return
        values = value if isinstance(value, list) else [value]
        field_map = self._index.get(field)
        if not field_map:
            return
        for v in values:
            key = _normalise(v)
            id_set = field_map.get(key)
            if id_set:
                id_set.discard(record_id)
                if not id_set:
                    del field_map[key]
        if not field_map:
            del self._index[field]

    # ── Lookup ─────────────────────────────────────────────────────────────────

    def eq(self, field: str, value: Any) -> set[str]:
        return set(self._index.get(field, {}).get(_normalise(value), set()))

    def neq(self, field: str, value: Any) -> set[str]:
        excluded = _normalise(value)
        result: set[str] = set()
        for key, ids in self._index.get(field, {}).items():
            if key != excluded:
                result.update(ids)
        return result

    def find_in(self, field: str, values: list[Any]) -> set[str]:
        result: set[str] = set()
        for v in values:
            result.update(self.eq(field, v))
        return result

    def not_in(self, field: str, values: list[Any]) -> set[str]:
        excluded = {_normalise(v) for v in values}
        result: set[str] = set()
        for key, ids in self._index.get(field, {}).items():
            if key not in excluded:
                result.update(ids)
        return result

    # ── Lifecycle ──────────────────────────────────────────────────────────────

    def clear(self) -> None:
        self._index.clear()

    @property
    def entry_count(self) -> int:
        return sum(len(m) for m in self._index.values())


def _normalise(value: Any) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value).lower()

File: rust/exact.rs
// =============================================================================
// vapor-idx — indexes/exact.rs
// Equality index: field → normalised_value → HashSet<record_id>
// Supports: eq, neq, in, not_in
// =============================================================================

use std::collections::{HashMap, HashSet};

#[derive(Debug, Default)]
pub struct ExactIndex {
    // field → normalised_value → set<id>
    index: HashMap<String, HashMap<String, HashSet<String>>>,
}

impl ExactIndex {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add(&mut self, field: &str, value: &serde_json::Value, id: &str) {
        match value {
            serde_json::Value::Array(items) => {
                for item in items {
                    self.add_single(field, &normalise(item), id);
                }
            }
            v => {
                self.add_single(field, &normalise(v), id);
            }
        }
    }

    fn add_single(&mut self, field: &str, key: &str, id: &str) {
        self.index
            .entry(field.to_owned())
            .or_default()
            .entry(key.to_owned())
            .or_default()
            .insert(id.to_owned());
    }

    pub fn remove(&mut self, field: &str, value: &serde_json::Value, id: &str) {
        match value {
            serde_json::Value::Array(items) => {
                for item in items {
                    self.remove_single(field, &normalise(item), id);
                }
            }
            v => self.remove_single(field, &normalise(v), id),
        }
    }

    fn remove_single(&mut self, field: &str, key: &str, id: &str) {
        if let Some(field_map) = self.index.get_mut(field) {
            if let Some(id_set) = field_map.get_mut(key) {
                id_set.remove(id);
                if id_set.is_empty() {
                    field_map.remove(key);
                }
            }
            if field_map.is_empty() {
                self.index.remove(field);
            }
        }
    }

    pub fn eq(&self, field: &str, value: &serde_json::Value) -> HashSet<String> {
        let key = normalise(value);
        self.index
            .get(field)
            .and_then(|m| m.get(&key))
            .cloned()
            .unwrap_or_default()
    }

    pub fn neq(&self, field: &str, value: &serde_json::Value) -> HashSet<String> {
        let excluded = normalise(value);
        let mut result = HashSet::new();
        if let Some(field_map) = self.index.get(field) {
            for (k, ids) in field_map {
                if k != &excluded {
                    result.extend(ids.iter().cloned());
                }
            }
        }
        result
    }

    pub fn find_in(&self, field: &str, values: &[serde_json::Value]) -> HashSet<String> {
        let mut result = HashSet::new();
        for v in values {
            result.extend(self.eq(field, v));
        }
        result
    }

    pub fn not_in(&self, field: &str, values: &[serde_json::Value]) -> HashSet<String> {
        let excluded: HashSet<String> = values.iter().map(normalise).collect();
        let mut result = HashSet::new();
        if let Some(field_map) = self.index.get(field) {
            for (k, ids) in field_map {
                if !excluded.contains(k) {
                    result.extend(ids.iter().cloned());
                }
            }
        }
        result
    }

    pub fn clear(&mut self) {
        self.index.clear();
    }

    pub fn entry_count(&self) -> usize {
        self.index.values().map(|m| m.len()).sum()
    }
}

fn normalise(value: &serde_json::Value) -> String {
    match value {
        serde_json::Value::String(s) => s.to_lowercase(),
        serde_json::Value::Bool(b)   => if *b { "true".into() } else { "false".into() },
        v                            => v.to_string(),
    }
}
